### PR TITLE
refactor(core): add `createMigrationCompilerHost`

### PR DESCRIPTION
**Pull request by alan-agius4**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
Current we need to create and override certain compiler host methods in every schematic because schematics use a virtual fs. We this change we extract this logic to a common util.

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: packages/core/schematics/utils/typescript/compiler_host.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  // if we run the migration for multiple tsconfig files which have intersecting
+  // source files, it can end up updating query definitions multiple times.
+  host.readFile = fileName => {
+    const buffer = tree.read(relative(normalize(basePath), normalize(fileName)));
+    // Strip BOM as otherwise TSC methods (Ex: getWidth) will return an offset which
+    // which breaks the CLI UpdateRecorder.
+    // See: https://github.com/angular/angular/pull/30719
+    return buffer ? buffer.toString().replace(/^\uFEFF/, '') : undefined;
+  };
+
```
Comment (alan-agius4) [inline packages/core/schematics/utils/typescript/compiler_host.ts]: @devversion as per our offline chat you should be able to mock `readResource` to an empty string here

File: packages/core/schematics/utils/typescript/compiler_host.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {Tree} from '@angular-devkit/schematics';
+import {CompilerHost, createCompilerHost} from '@angular/compiler-cli';
+import {relative} from 'path';
+import * as ts from 'typescript';
+
+export function createMigrationCompilerHost(
+    tree: Tree, options: ts.CompilerOptions, basePath: string): CompilerHost {
+  const host = createCompilerHost({options});
```
Comment (devversion) [inline packages/core/schematics/utils/typescript/compiler_host.ts]: I'm confused. why do we want to always use the `compiler-cli` migration host here? Only the `undecorated-classes-with-di` and `static-query` migration need to wire up a NGC program, so for those we could just pass in the `ts.CompilerHost` when we create the `AngularCompilerProgram`.

I think we should just use `ts.CompilerHost` here and remove the dep on `compiler-cli` in the utils.

Comment (alan-agius4) [reply packages/core/schematics/utils/typescript/compiler_host.ts]: PTAL

File: packages/core/schematics/migrations/undecorated-classes-with-di/index.ts [hunk 3]
```diff
@@ -147,19 +149,8 @@ function gracefullyCreateProgram(
...
-        const buffer = tree.read(relative(basePath, fileName));
-        // Strip BOM as otherwise TSC methods (Ex: getWidth) will return an offset which
-        // which breaks the CLI UpdateRecorder.
-        // See: https://github.com/angular/angular/pull/30719
-        return buffer ? buffer.toString().replace(/^\uFEFF/, '') : undefined;
-      };
-
-      return host;
+      const tsHost = createMigrationCompilerHost(tree, options, basePath);
+      return createCompilerHost({options, tsHost});
```
Comment (devversion) [inline packages/core/schematics/migrations/undecorated-classes-with-di/index.ts]: just like before, we should return the `tsHost` here? no need to create the `compiler-cli` host from it.

**Review by devversion — APPROVED** [review]
LGTM, one minor comment in the `undecorated-classes-with-di` migration.

File: packages/core/schematics/utils/typescript/compiler_host.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import * as ts from 'typescript';
+
+export function createMigrationCompilerHost(
+    tree: Tree, options: ts.CompilerOptions, basePath: string): ts.CompilerHost {
+  const host = ts.createCompilerHost(options, true);
+
+  // We need to overwrite the host "readFile" method, as we want the TypeScript
+  // program to be based on the file contents in the virtual file tree. Otherwise
+  // if we run the migration for multiple tsconfig files which have intersecting
+  // source files, it can end up updating query definitions multiple times.
```
Comment (devversion) [inline packages/core/schematics/utils/typescript/compiler_host.ts]: The comment here is specific to the static-query migration. Maybe remove the mentioning of `query definitions` here?

Comment (alan-agius4) [reply packages/core/schematics/utils/typescript/compiler_host.ts]: done

File: packages/core/schematics/migrations/undecorated-classes-with-di/index.ts (hunk 3 above)
Comment (alan-agius4) [reply packages/core/schematics/migrations/undecorated-classes-with-di/index.ts]: done

**Review by devversion — APPROVED** [review]
LGTM

**Review by kara — APPROVED** [review]
LGTM, thanks for the cleanup!

**Comment by kara** [comment]
[sanity presubmit](http://test/OCL:272656690:BASE:272656717:1570112959242:c6847db)

**Comment by kara** [comment]
Caretaker note: Needs g3 patch (can use [this](https://critique.corp.google.com/#review/272666466))

**Pull request closed without merging**